frame/header-mmr/src/primitives.rs

```diff
@@ -0,0 +1,192 @@
+pub use mmr::MMR;
+
+// --- crates.io ---
+use codec::{Decode, Encode};
+// --- github.com ---
+use mmr::{Error, MMRStore, Merge, MerkleProof, Result as MMRResult};
+// --- paritytech ---
+use sp_core::offchain::StorageKind;
+use sp_io::{offchain, offchain_index};
+use sp_runtime::traits::Hash;
+use sp_std::{marker::PhantomData, prelude::*};
+// --- darwinia ---
+use crate::*;
+
+pub struct Hasher<T>(PhantomData<T>);
+impl<T: Config> Merge for Hasher<T> {
+	type Item = T::Hash;
+
+	fn merge(lhs: &Self::Item, rhs: &Self::Item) -> Self::Item {
+		T::Hashing::hash_of(&(lhs, rhs))
+	}
+}
+
+pub struct OffchainStorage;
+pub struct RuntimeStorage;
+pub struct Storage<StorageType, T>(PhantomData<(StorageType, T)>);
+impl<StorageType, T> Default for Storage<StorageType, T> {
+	fn default() -> Self {
+		Self(Default::default())
+	}
+}
+impl<T> MMRStore<T::Hash> for Storage<OffchainStorage, T>
+where
+	T: Config,
+{
+	fn get_elem(&self, position: NodeIndex) -> MMRResult<Option<T::Hash>> {
+		let key = <Pallet<T>>::offchain_key(position);
+
+		if let Some(v) = offchain::local_storage_get(StorageKind::PERSISTENT, &key) {
+			Ok(Decode::decode(&mut &*v).ok())
+		} else {
+			Ok(<MMRNodeList<T>>::get(position))
+		}
+
+		// TODO: once the pruning finish, restore this
+		// Ok(offchain::local_storage_get(StorageKind::PERSISTENT, &key)
+		// .and_then(|v| Decode::decode(&mut &*v).ok()))
+	}
+
+	fn append(&mut self, _: NodeIndex, _: Vec<T::Hash>) -> MMRResult<()> {
+		log::error!("Not allow to append elem(s) in the off-chain context!");
+
+		Err(Error::InconsistentStore)
+	}
+}
+impl<T> MMRStore<T::Hash> for Storage<RuntimeStorage, T>
+where
+	T: Config,
+{
+	fn get_elem(&self, position: NodeIndex) -> MMRResult<Option<T::Hash>> {
+		Ok(<Pallet<T>>::peak_of(position))
+	}
+
+	// ? Actually we can move this pruning logic outside
+	// ? and perform the pruning every X blocks
+	// ?
+	// ? commented by Xavier
+	fn append(&mut self, position: NodeIndex, elems: Vec<T::Hash>) -> MMRResult<()> {
```

Comment:
Since this is an internal function for MMRStore which in theory can append elems that is not generated by add one leaf.

Considering the using context in the pallet & runtime, there might not be exceptions, but pruning here is not match the function api ideally.